src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1451,19 +1476,45 @@ protected override BoundNode VisitReturnStatementNoAdjust(BoundReturnStatement n
             if (_returnTypesOpt == null &&
                 TryGetReturnType(out TypeWithAnnotations returnType, out FlowAnalysisAnnotations returnAnnotations))
             {
-                TypeWithState returnState;
-                if (node.RefKind == RefKind.None)
+                if (node.RefKind == RefKind.None &&
+                    returnType.Type.SpecialType == SpecialType.System_Boolean &&
+                    isUnconvertedBooleanExpression(expr))
```

Comment:
Why do we need to look at the returned expression before conversions are applied?

Reply:
If there is a conversion from some custom type to bool, then we can't do the new analysis. We don't know what logic that conversion might contain. We lose track of whenTrue/False branches in conversions, since it's custom logic.

Also imagine if that conversion produced nullability warnings, we'd want to run it through `VisitImplicitConversion`.